### PR TITLE
Added `use_path_derivative` as an option to all re-parameterizable distributions

### PR DESCRIPTION
**Pull request by botev**
Addressing #67

### CODE REVIEW

**Comment by botev** [comment]
I did not manage to run the tests as I don't have tf on my travelling notebook, so let's see if they pass.

File: tests/distributions/test_multivariate.py [hunk 1]
```diff
@@ -456,6 +456,33 @@ def test_sample_reparameterized(self):
         self.assertEqual(t_grads, None)
         self.assertEqual(logits_grads, None)
 
+    def test_path_derivative(self):
+        temperature = tf.ones([])
```
Comment (thjashin) [inline tests/distributions/test_multivariate.py]: Maybe just pass `1`?

File: tests/distributions/test_multivariate.py [hunk 2]
```diff
@@ -456,6 +456,33 @@ def test_sample_reparameterized(self):
...
+        logits = tf.ones([2, 3])
+        n_samples = tf.placeholder(tf.int32, shape=[])
+
+        con_rep = ExpConcrete(temperature, logits, use_path_derivative=True)
+        samples = con_rep.sample(n_samples)
+        log_prob = con_rep.log_prob(samples)
+        t_path_grads, logits_path_grads = tf.gradients(log_prob, [temperature, logits])
+        sample_grads = tf.gradients(log_prob, samples)
+        t_true_grads = tf.gradients(samples, temperature, sample_grads)[0]
+        logits_true_grads = tf.gradients(samples, logits, sample_grads)[0]
```
Comment (thjashin) [inline tests/distributions/test_multivariate.py]: Nice use of `tf.gradients`. Didn't notice it supports initial gradients before.

**Comment by thjashin** [comment]
@botev Sorry for being late. I've worked through the changes and there seems to be no problems. Only minor comments.

File: tests/distributions/test_multivariate.py (hunk 2 above)
Comment (botev) [reply tests/distributions/test_multivariate.py]: Yes, in fact `tf.gradients` is equivalent to `theano.tensor.Lop`.

File: tests/distributions/test_multivariate.py (hunk 1 above)
Comment (botev) [reply tests/distributions/test_multivariate.py]: I'll update that now.

**Comment by botev** [comment]
@thjashin I've changed to `tf.constant(1.0)` but I need it to be a tensor variable so the `tf.gradients` is valid operation.

**Comment by thjashin** [comment]
Cool. I'll merge. Thanks for contributing. That's a lot of work!
